### DISCUSSION TITLE
How to add Event Script to HFS 0.53.0

### DISCUSSION THREAD

**Discussion by LinYouMin** [rejetto/hfs#746] [General]
Dear Sirs,

I'm new to HFS.
My current version is 0.53.0 on MacOS.
I can not find a Menu item of `Menu > Other options > Edit event scripts` in the HFS Admin-panel.

I'd like to add Event Script to the HFS like to add a new directory when get the request.

[request]
{.if|{.match|/create-directory/*|%url.}|{
{.mkdir|%url-folder%.}
{.reply|200|Directory created.}
}.}

Please advise how I can accomplish this simple task.

Thanks!

Scott Lin

**Comment by rejetto** [comment]
in new HFS this is different.
the correct way to intercept a request is to write a middleware.
Find documentation at https://github.com/rejetto/hfs/wiki/Middlewares

but you DON'T NEED to make "/create-directory", as you already have that in HFS. It's `/~/api/create_folder`.
You can see it in action if you have upload permission.
You can give permission to an account, or to anyone, it's up to you.
When you have permission, you'll see a new "upload" button, and if you click it you'll see "create folder".

**Reply by LinYouMin** [reply]
Thanks!

I'm trying to implement a request to the create_folder api from python Django as:

# Send the POST reuqest to create_folder api
response = requests.post('create_folder', hfs_url, folder=folder)

hfs_url is: hfs_url: http://192.168.1.106/hfs/PX001/2634
folder is: 2024-09-22

However, I got an error message:

TypeError: Session.request() got an unexpected keyword argument 'folder'

Please advise how to correct the requests to this create_folder api.

Thanks!

**Comment by rejetto** [comment]
your problem now is a python problem, it has nothing to do with HFS, and i'm no good with python.
what i can understand is that you are passing wrong parameters to post function.

**Comment by LinYouMin** [comment]
Thanks!

Would you please advise what specific arguments are required for this create_folder api. I can not find the document for this api.

**Comment by rejetto** [comment]
i just documented it
https://hfs-3.apidog.io/api-10249487

**Reply by LinYouMin** [reply]
Dear Sirs,

Thanks for your prompt hard working!

I revise my function and got a "bad uri" error message.
Following is the Link, Python and Error message.

# Link on the HFS Admin Panel

**link = http://192.1.168.106/hfs/2513/**

# Python

def hfs_mkdir(scheme, ip_addr, hfs_path, folder_name):

    api_path = os.path.join('~', 'api', 'create_folder')
    hfs_url = make_upload_url(scheme, ip_addr, api_path)


    params = {}
    params['uri'] = f'/{hfs_path}/'  # Ensure the URI is properly formatted
    params['name'] = folder_name

    print("hfs_url:", hfs_url)
    print("params:", params)

    # Send the POST reuqest to create_folder api
    response = requests.post(hfs_url, params=params)


# Error message

(django) youmin@YouMindgdiannao ftp_client % python ftp_client.py

hfs_url: **http://192.168.1.106/~/api/create_folder**
params: **{'uri': '/hfs/2513/', 'name': '2024-09-23'}**
Failed to create folder. Status code: 400
Response: bad uri

Please advise how should I construct the "uri" parameter.

Scott Lin

**Comment by LinYouMin** [comment]
Revise the python request with additonal authentication, but still got a bad uri response error message.

# Python

    auth = HTTPBasicAuth(username, password)

    # Send the POST reuqest to create_folder api
    response = requests.post(hfs_url, params=params, auth=auth)

# Error Message

hfs_url: http://192.168.1.106/~/api/create_folder
params: {'uri': '/hfs/2513/', 'name': '2024-09-23'}
Failed to create folder. Status code: 400
Response: bad uri

**Comment by rejetto** [comment]
i see you are passing "params" to requests.post.
i googled for requests.post documentation and i don't see "params".
I only see params for get, not for post.
your problem is a python problem.

**Comment by LinYouMin** [comment]
Dear Sirs,

See below the information.
The following python Django code works now.

Change log:
1. 'uri': '/hfs/PX001/2513/'
2. response = requests.post(hfs_url, json=params, auth=auth)

Your kindly support is highly appreciated.

Scott Lin


# Link on the HFS Admin Panel

Disk source = /Users/youmin/hfs/PX001/0646
link = http://192.1.168.106/hfs/2513/

# Python

def hfs_mkdir(scheme, ip_addr, hfs_path, folder_name, username, password):

    api_path = os.path.join('~', 'api', 'create_folder')
    hfs_url = make_upload_url(scheme, ip_addr, api_path)

    params = {}
    params['uri'] = f'/{hfs_path}/'  # Ensure the URI is properly formatted
    params['name'] = folder_name

    print("hfs_url:", hfs_url)
    print("params:", params)

    auth = HTTPBasicAuth(username, password)

    # Send the POST reuqest to create_folder api
    response = requests.post(hfs_url, json=params, auth=auth)

    # Check the server's response
    if response.status_code == 200:
        print("Create folder successfully!")
        message = "Create folder successfully!"
        status = True
    else:
        print(f"Failed to create folder. Status code: {response.status_code}")
        print(f"Response: {response.text}")
        message = response.text
        status = False

    return status, message

# Error message

(django) youmin@YouMindgdiannao ftp_client % python ftp_client.py

hfs_url: http://192.168.1.106/~/api/create_folder
params: {'uri': '/hfs/PX001/2513/', 'name': '2024-09-23'}
Create folder successfully!

**Comment by rejetto** [comment]
happy to know it's working!
